### PR TITLE
Enable DINO to OTX - Step 1. Enable Deformable DETR to OTX

### PR DESCRIPTION
**Pull request by jaegukhyun**
### Summary
As a first step to bring DINO to OTX, this PR enable Deformable DETR to OTX.
Since mmdeploy do not support Deformable DETR(it supports only DETR), there are some custom patches for mmcv and mmdeploy ops

**Known Issue**
1. Current Deformable DETR use fixed schedule, not early stopping and plateau lr decay 
2. Current Deformable DETR does not support smart weight loading, and other class incremental technique
3. Current DetClassProbabilityMapHook does not support End-to-End detector
4. When inference with exported Deformable DETR, there are some bottleneck for some images
image

These known issues are handled after implementing DINO from this PR

### How to test
Unit tests and integration tests are added. E2E Test will be added after implementing DINO

### Checklist

- [x] I have added unit tests to cover my changes.​
- [x] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [x] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [x] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [x] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: otx/algorithms/common/adapters/mmcv/ops/multi_scale_deformable_attn_pytorch.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+def multi_scale_deformable_attn_pytorch(
+    value: torch.Tensor,
+    value_spatial_shapes: torch.Tensor,
+    sampling_locations: torch.Tensor,
+    attention_weights: torch.Tensor,
+) -> torch.Tensor:
+    """Custom patch for multi_scale_deformable_attn_pytorch function.
+
+    Original implementation in mmcv.ops use torch.nn.functional.grid_sample.
+    It occurs errors during inference with OpenVINO exported model.
```
Comment (eunwoosh) [inline otx/algorithms/common/adapters/mmcv/ops/multi_scale_deformable_attn_pytorch.py]:
```suggestion
    It makes errors during inference with OpenVINO exported model.
```

File: otx/algorithms/common/adapters/mmcv/ops/multi_scale_deformable_attn_pytorch.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    Therefore this function change grid_sample function to custom mmcv_grid_sample function.
```
Comment (eunwoosh) [inline otx/algorithms/common/adapters/mmcv/ops/multi_scale_deformable_attn_pytorch.py]: Is this function same except using `_custom_grid_sample`?
If then, I think writing that information in comments can makes code maintenance easier later.

File: otx/algorithms/detection/adapters/mmdet/models/detectors/custom_deformable_detr_detector.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+
+if is_mmdeploy_enabled():
+    from mmdeploy.core import FUNCTION_REWRITER
+
+    @FUNCTION_REWRITER.register_rewriter(
+        "otx.algorithms.detection.adapters.mmdet.models.detectors.custom_deformable_detr_detector.CustomDeformableDETR.simple_test"
+    )
+    def custom_deformable_detr__simple_test(ctx, self, img, img_metas, **kwargs):
+        """Function for custom_mask_rcnn__simple_test."""
```
Comment (eunwoosh) [inline otx/algorithms/detection/adapters/mmdet/models/detectors/custom_deformable_detr_detector.py]:
```suggestion
        """Function for custom_deformable_det__simple_test."""
```
I think it's typo.

File: otx/algorithms/detection/configs/detection/resnet50_deformable-detr/data_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
+"""Data pipeline for Deformable DETR."""
+# dataset settings
+dataset_type = "CocoDataset"
+data_root = "data/coco/"
+img_norm_cfg = dict(mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
+
+# train_pipeline, NOTE the img_scale and the Pad's size_divisor is different
```
Comment (eunwoosh) [inline otx/algorithms/detection/configs/detection/resnet50_deformable-detr/data_pipeline.py:7]: Could you explain why it is different?

File: otx/algorithms/detection/configs/detection/resnet50_deformable-detr/template_experimental.yaml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+      batch_size:
+        default_value: 2
+        auto_hpo_state: POSSIBLE
+      learning_rate:
+        default_value: 0.0002
+        auto_hpo_state: POSSIBLE
+      learning_rate_warmup_iters:
+        default_value: 3
+      num_iters:
+        default_value: 12
```
Comment (eunwoosh) [inline otx/algorithms/detection/configs/detection/resnet50_deformable-detr/template_experimental.yaml:37]: It's just a question. Is there any reason why it's different from models?

File: otx/algorithms/detection/configs/detection/resnet50_deformable-detr/template_experimental.yaml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+# Training resources.
+max_nodes: 1
+training_targets:
+  - GPU
+  - CPU
+
+# Stats.
+gigaflops: ???
+size: ???
+# # Inference options. Defined by OpenVINO capabilities, not Algo Backend or Platform.
```
Comment (eunwoosh) [inline otx/algorithms/detection/configs/detection/resnet50_deformable-detr/template_experimental.yaml:60]: Can this part be removed?

File: otx/algorithms/detection/configs/detection/resnet50_deformable-detr/template_experimental.yaml [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        default_value: Incremental
+
+# Training resources.
+max_nodes: 1
+training_targets:
+  - GPU
+  - CPU
+
+# Stats.
+gigaflops: ???
```
Comment (eunwoosh) [inline otx/algorithms/detection/configs/detection/resnet50_deformable-detr/template_experimental.yaml:58]: Could you fill this information?

File: tests/unit/algorithms/common/adapters/mmdeploy/ops/test_custom_ops.py
Comment (eunwoosh) [inline tests/unit/algorithms/common/adapters/mmdeploy/ops/test_custom_ops.py:1]: Could you add unit test for functions?

**Review by eunwoosh — COMMENTED** [review]
Thanks for adding great model :)
I left some comments. please take a look.

File: otx/algorithms/common/adapters/mmcv/ops/multi_scale_deformable_attn_pytorch.py (hunk 1 above)
Comment (harimkang) [reply otx/algorithms/common/adapters/mmcv/ops/multi_scale_deformable_attn_pytorch.py]: how about `raise`

File: otx/algorithms/common/adapters/mmcv/ops/multi_scale_deformable_attn_pytorch.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+def multi_scale_deformable_attn_pytorch(
+    value: torch.Tensor,
+    value_spatial_shapes: torch.Tensor,
+    sampling_locations: torch.Tensor,
+    attention_weights: torch.Tensor,
+) -> torch.Tensor:
+    """Custom patch for multi_scale_deformable_attn_pytorch function.
+
+    Original implementation in mmcv.ops use torch.nn.functional.grid_sample.
```
Comment (harimkang) [inline otx/algorithms/common/adapters/mmcv/ops/multi_scale_deformable_attn_pytorch.py:20]: What about adding an address from an existing implementation?

File: otx/algorithms/common/adapters/mmdeploy/ops/custom_ops.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+from torch.onnx import symbolic_helper
+from torch.onnx._internal import jit_utils
+
+# Remove previous registered symbolic
+SYMBOLIC_REWRITER._registry._rewrite_records["squeeze"] = list()
+SYMBOLIC_REWRITER._registry._rewrite_records["grid_sampler"] = list()
+
+
+@SYMBOLIC_REWRITER.register_symbolic("squeeze", is_pytorch=True)
+def squeeze__default(ctx, g, self, dim=None):
```
Comment (harimkang) [inline otx/algorithms/common/adapters/mmdeploy/ops/custom_ops.py:16]: What does the double underbar mean?

File: otx/algorithms/detection/adapters/mmdet/models/detectors/custom_deformable_detr_detector.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+
+@DETECTORS.register_module()
+class CustomDeformableDETR(DeformableDETR):
+    """Custom Deformable DETR with task adapt.
+
+    Deformable DETR does not support task adapt, so it just take task_adpat argument.
+    """
+
+    def __init__(self, *args, task_adapt=None, **kwargs):
```
Comment (harimkang) [inline otx/algorithms/detection/adapters/mmdet/models/detectors/custom_deformable_detr_detector.py:24]: We'll have to think about removing task_adapt in the future. This comment is unrelated to this PR. :)

**Review by harimkang — COMMENTED** [review]
Welcome to the new algorithm. It looks good for the most part, and I have a few comments.

File: otx/algorithms/common/adapters/mmcv/ops/__init__.py [hunk 11]
```diff
@@ -0,0 +1,10 @@
+"""Initial file for mmcv ops."""
+# Copyright (C) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
+
+from mmcv.ops import multi_scale_deform_attn
+
+from .multi_scale_deformable_attn_pytorch import multi_scale_deformable_attn_pytorch
+
+multi_scale_deform_attn.multi_scale_deformable_attn_pytorch = multi_scale_deformable_attn_pytorch
```
Comment (sungmanc) [inline otx/algorithms/common/adapters/mmcv/ops/__init__.py]: Patching the function in `__init__.py` is not common. How about moving this patching to the `adapters/mmcv/ops/multi_scale_deformable_attn_pytorch.py`

File: otx/algorithms/detection/configs/detection/resnet50_deformable-detr/data_pipeline.py (hunk 4 above)
Comment (jaegukhyun) [reply otx/algorithms/detection/configs/detection/resnet50_deformable-detr/data_pipeline.py:7]: I hope I can explain this, but I just brought data_pipeline.py from mmdetection.

File: otx/algorithms/detection/configs/detection/resnet50_deformable-detr/template_experimental.yaml (hunk 5 above)
Comment (jaegukhyun) [reply otx/algorithms/detection/configs/detection/resnet50_deformable-detr/template_experimental.yaml:37]: If you ask why deformable detr is different with other detectors(ATSS, SSD), the answer is current deformable detr does not support ReducedLRPlateau decay. The LR scheduler don't works properly. I'll modify this in another PR.

If you ask why template.yaml is different from model.py, then the answer also same. It will be modified in another PR.

File: otx/algorithms/detection/configs/detection/resnet50_deformable-detr/template_experimental.yaml (hunk 6 above)
Comment (jaegukhyun) [reply otx/algorithms/detection/configs/detection/resnet50_deformable-detr/template_experimental.yaml:60]: I'm not sure, I just brought this from other model's template. What do you think?

File: otx/algorithms/detection/configs/detection/resnet50_deformable-detr/template_experimental.yaml (hunk 7 above)
Comment (jaegukhyun) [reply otx/algorithms/detection/configs/detection/resnet50_deformable-detr/template_experimental.yaml:58]: This information will be added another PR. That PR will change template_experimental.yaml to template.yaml

File: otx/algorithms/common/adapters/mmdeploy/ops/custom_ops.py (hunk 9 above)
Comment (jaegukhyun) [reply otx/algorithms/common/adapters/mmdeploy/ops/custom_ops.py:16]: I don't know... I think it is convention from mmdeploy.

File: otx/algorithms/detection/adapters/mmdet/models/detectors/custom_deformable_detr_detector.py (hunk 10 above)
Comment (jaegukhyun) [reply otx/algorithms/detection/adapters/mmdet/models/detectors/custom_deformable_detr_detector.py:24]: I really hope so

File: otx/algorithms/detection/configs/detection/resnet50_deformable-detr/template_experimental.yaml (hunk 5 above)
Comment (eunwoosh) [reply otx/algorithms/detection/configs/detection/resnet50_deformable-detr/template_experimental.yaml:37]: Sorry for my ambiguous. I just want to ask about `num_iters` default value.

File: otx/algorithms/detection/configs/detection/resnet50_deformable-detr/template_experimental.yaml (hunk 6 above)
Comment (eunwoosh) [reply otx/algorithms/detection/configs/detection/resnet50_deformable-detr/template_experimental.yaml:60]: Maybe we need to remove this part including other model's template. For now, let it be.

File: otx/algorithms/detection/configs/detection/resnet50_deformable-detr/template_experimental.yaml (hunk 5 above)
Comment (jaegukhyun) [reply otx/algorithms/detection/configs/detection/resnet50_deformable-detr/template_experimental.yaml:37]: Current Deformable DETR does not support earlystopping, same reason with ReducedLRPlateau. I'll modify this in another PR.

File: otx/algorithms/common/adapters/mmcv/ops/multi_scale_deformable_attn_pytorch.py (hunk 1 above)
Comment (jaegukhyun) [reply otx/algorithms/common/adapters/mmcv/ops/multi_scale_deformable_attn_pytorch.py]: Changed to raise

File: otx/algorithms/common/adapters/mmcv/ops/multi_scale_deformable_attn_pytorch.py (hunk 2 above)
Comment (jaegukhyun) [reply otx/algorithms/common/adapters/mmcv/ops/multi_scale_deformable_attn_pytorch.py]: Changed from custom mmcv_grid_sample -> _custom_grid_sample

File: otx/algorithms/detection/adapters/mmdet/models/detectors/custom_deformable_detr_detector.py (hunk 3 above)
Comment (jaegukhyun) [reply otx/algorithms/detection/adapters/mmdet/models/detectors/custom_deformable_detr_detector.py]: Changed Thanks

File: tests/unit/algorithms/common/adapters/mmdeploy/ops/test_custom_ops.py
Comment (jaegukhyun) [reply tests/unit/algorithms/common/adapters/mmdeploy/ops/test_custom_ops.py:1]: Added. Thanks

File: otx/algorithms/common/adapters/mmcv/ops/multi_scale_deformable_attn_pytorch.py (hunk 8 above)
Comment (jaegukhyun) [reply otx/algorithms/common/adapters/mmcv/ops/multi_scale_deformable_attn_pytorch.py:20]: I added function address which replace nn.functional.grid_sample

File: otx/algorithms/common/adapters/mmcv/ops/__init__.py (hunk 11 above)
Comment (jaegukhyun) [reply otx/algorithms/common/adapters/mmcv/ops/__init__.py]: Changed Thanks

**Review by sungmanc — APPROVED** [review]

**Review by eunwoosh — APPROVED** [review]

**Review by harimkang — APPROVED** [review]
LGTM!

BTW, Anything to add to the documentation for this?

**Comment by jaegukhyun** [comment]
> LGTM!
> 
> BTW, Anything to add to the documentation for this?

Documentation will be updated after adding DINO and conducting some experiments on public datasets
